src/panels/lovelace/cards/hui-shopping-list-card.ts

```diff
@@ -0,0 +1,229 @@
+import { html, LitElement } from "@polymer/lit-element";
+import { repeat } from "lit-html/directives/repeat";
+import "@polymer/paper-checkbox/paper-checkbox.js";
+import "@polymer/paper-icon-button/paper-icon-button.js";
+import "@polymer/paper-input/paper-input.js";
+import "@polymer/paper-item/paper-icon-item.js";
+import "@polymer/paper-item/paper-item-body.js";
+
+import "../../../components/ha-card.js";
+
+import { hassLocalizeLitMixin } from "../../../mixins/lit-localize-mixin";
+import { HomeAssistant } from "../../../types.js";
+import { LovelaceCard, LovelaceConfig } from "../types.js";
+import {
+  clearItems,
+  fetchItems,
+  addItem,
+  completeItem,
+  saveEdit,
+} from "../../../data/shopping_list";
+import { TemplateResult } from "lit-html";
+
+interface Config extends LovelaceConfig {
+  title?: string;
+}
+
+export interface ShoppingListItem {
+  id: number;
+  name: string;
+  complete: boolean;
+}
+
+class HuiShoppingListCard extends hassLocalizeLitMixin(LitElement)
+  implements LovelaceCard {
+  public hass?: HomeAssistant;
+  protected config?: Config;
+  private items?: ShoppingListItem[];
+  private _unsubEvents?: () => Promise<void>;
+
+  static get properties() {
+    return {
+      config: {},
+      items: {},
+    };
+  }
+
+  public getCardSize(): number {
+    return (
+      (this.config ? (this.config.title ? 1 : 0) : 0) +
+      (this.items ? this.items.length : 3)
+    );
+  }
+
+  public setConfig(config: Config): void {
+    this.config = config;
+    this.items = [];
+    this._fetchData();
+  }
+
+  public connectedCallback() {
+    if (this.hass) {
+      super.connectedCallback();
+      this.hass.connection
+        .subscribeEvents(this._fetchData, "shopping_list_updated")
+        .then(function(this: HuiShoppingListCard, unsub: () => Promise<void>) {
```

Comment:
the reason `this` is undefined is because you're using `function` instead of a lambda (`=>` syntax). You should always prefer the arrow syntax, it has less overhead and it maintains the `this` of the context in which it is defined.

Reply:
are you sure? Can you make it `() => this._fetchData()` instead of passing in `this._fetchData` ?

Reply:
Thanks! I don't think I would have figured that one out. works now!